File: 专题/状态DP/股票问题/122.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp = [ [0] * (n+1) for i in range(2)]
        dp[1][0] = - 100000 # prices[i] <= e4
        for i in range(n):
            dp[0][i+1] = max(dp[1][i] + prices[i], dp[0][i])
            dp[1][i+1] = max(dp[1][i], dp[0][i] - prices[i])
        return dp[0][n]

    def maxProfit2(self, prices: List[int]) -> int:
        n = len(prices)
        empty = 0
        exist = -10000000000
        for i in range(n):
            temp = empty
            empty = max(exist + prices[i], empty)
            exist = max(exist, temp - prices[i])
        return empty

File: 专题/状态DP/股票问题/122.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int empty[n+1], exist[n+1];
        empty[0] = 0;
        exist[0] = INT_MIN;
        for (int i = 0; i < n; ++i) {
            empty[i+1] = max(empty[i], exist[i] + prices[i]);
            exist[i+1] = max(exist[i], empty[i] - prices[i]);
        }
        return empty[n];
    }
    int maxProfit2(vector<int>& prices) {
        int n = prices.size();
        int empty = 0, exist = INT_MIN, temp;   
        for (int i = 0; i < n; ++i) {
            temp = empty;
            empty = max(empty, exist + prices[i]);
            exist = max(exist, temp - prices[i]);
        }
        return empty;
    }
};